packages/spotlight/src/server/utils/cors.ts

```diff
@@ -1,15 +1,273 @@
+      }
+    }
+  }
+
+  // Process OS resolver results (no TTL available - from /etc/hosts or system DNS cache)
+  if (lookupResult.status === "fulfilled") {
+    const lookupAddresses = Array.isArray(lookupResult.value) ? lookupResult.value : [lookupResult.value];
+    for (const record of lookupAddresses) {
+      // Avoid duplicates from DNS resolve
+      if (!results.some(r => r.address === record.address)) {
+        results.push({ address: record.address }); // No TTL from lookup
+      }
+    }
+  }
+
+  return { addresses: results, minTtl };
+}
+
+/**
+ * Calculate the cache TTL based on DNS response.
+ *
+ * TTL Rules:
+ * 1. If DNS provides TTL >= 6h: use DNS TTL
+ * 2. If DNS provides TTL < 6h: use 6h (minimum for rebinding protection)
+ * 3. If no TTL provided (e.g., /etc/hosts): use 1h default
+ */
+function calculateCacheTtl(minTtl?: number): number {
+  if (minTtl === undefined) {
+    return DEFAULT_TTL_MS;
+  }
+  // Convert seconds to milliseconds and enforce minimum
+  const ttlMs = minTtl * 1000;
+  return Math.max(ttlMs, MIN_TTL_MS);
+}
+
+/**
+ * Clear the DNS cache and machine IPs cache. Useful for testing.
+ */
+export function clearDnsCache(): void {
+  dnsCache.clear();
+  machineIPsCache = null;
+  machineIPsCacheTime = 0;
+}
+
+/**
+ * Get the current cache size. Useful for testing and monitoring.
+ */
+export function getDnsCacheSize(): number {
+  return dnsCache.size;
+}
+
 /**
  * Validates if an origin should be allowed to access the Sidecar.
  *
  * Allowed origins:
- * - localhost with any port or protocol (http/https)
+ * - Any origin whose hostname resolves to this machine's IPs:
+ *   - Loopback addresses (127.0.0.1, ::1)
+ *   - Any IP assigned to the machine's network interfaces
+ *     (local IPs, VPN IPs like Tailscale/Zerotier, etc.)
  * - https://spotlightjs.com (HTTPS only, default port)
  * - https://*.spotlightjs.com (HTTPS only, default port)
  *
+ * Security: DNS Rebinding Protection
+ * ==================================
+ * Instead of hardcoding "localhost" and known IPs, we resolve hostnames via DNS
+ * and check if they point to this machine. This allows:
+ * - Custom local hostnames (e.g., from /etc/hosts)
+ * - VPN-based access (Tailscale, Zerotier, etc.)
+ * - Custom domains pointing to the machine
+ *
+ * We protect against DNS rebinding attacks through a minimum 6-hour cache TTL.
+ *
+ * Threat Model:
+ * - An attacker controlling evil.com could try to rebind it to 127.0.0.1
+ * - Our 6h minimum cache means the attacker must keep the user engaged for 6+ hours
+ * - Even then, rebinding to 127.0.0.1 would make their site inaccessible
+ * - This makes DNS rebinding attacks impractical against Spotlight
+ *
+ * For machine IPs (non-loopback), the risk is even lower because:
+ * - External access requires explicit port forwarding/exposure
+ * - Most machines are behind NAT, blocking inbound connections by default
```

Comment:
This comment seems to be a repetition from above. Just refer to the parts above this place instead of repeating content.